{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# **E1. ACTIVIDAD INTEGRADORA 1**\n",
        "\n",
        "Rorigo Lopez Murguia\n",
        "\n",
        "Victor Misael Escalante Alvarado\n",
        "\n"
      ],
      "metadata": {
        "id": "9Lquji78WyBP"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Dependencias del proyecto\n",
        "\n",
        "Funciones utilizadas para el funcionamiento del proyecto (Es necesario ejecutar antes de las siguientes celdas)"
      ],
      "metadata": {
        "id": "-Dmv_jtJK5N4"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Funciones del proyecto\n",
        "\n",
        "class Protein:\n",
        "    def __init__(self, name, sequence):\n",
        "        self.name = name\n",
        "        self.sequence = sequence\n",
        "\n",
        "    def __str__(self):\n",
        "        return f\"{self.name}: {self.sequence}\"\n",
        "\n",
        "#Z array para string-matching\n",
        "def ZArray(texto, patron):\n",
        "    # Z - Array Algorithm\n",
        "    n = len(texto)\n",
        "    m = len(patron)\n",
        "\n",
        "    L = 0\n",
        "    R = 0\n",
        "    x = n + m + 1\n",
        "    Z = [-1 for i in range(x)]\n",
        "    C = patron + \"$\" + texto\n",
        "\n",
        "    for i in range(x):\n",
        "        if i <= R:\n",
        "            k = i - L\n",
        "            if Z[k] < R - i + 1:\n",
        "                Z[i] = Z[k]\n",
        "            else:\n",
        "                L = i\n",
        "                while R < x and C[R-L] == C[R]:\n",
        "                    R += 1\n",
        "                Z[i] = R - L\n",
        "                R -= 1\n",
        "        else:\n",
        "            L = R = i\n",
        "            while R < x and C[R-L] == C[R]:\n",
        "                R += 1\n",
        "            Z[i] = R - L\n",
        "            R -= 1\n",
        "    return Z\n",
        "\n",
        "def StringMatchZ(texto, patron):\n",
        "  m = len(texto)\n",
        "  n = len(patron)\n",
        "  m = m + n + 1\n",
        "  indexes = []\n",
        "  zArray = ZArray(texto,patron)\n",
        "  for i in range(m):\n",
        "    if zArray[i] == n:\n",
        "      # return i - n - 1\n",
        "      indexes.append(i-n-1)\n",
        "  return indexes\n",
        "\n",
        "#Manacher para palíndromo más largo en los genes\n",
        "def Manacher(frase):\n",
        "    centro = limite = 0\n",
        "\n",
        "    texto = \"@$\" + \"$\".join(frase) + \"$#\"\n",
        "    e = len(texto)\n",
        "\n",
        "    P = [0] * e\n",
        "\n",
        "    for i in range(1, e - 1):\n",
        "        if i < limite:\n",
        "            simetrica = 2 * centro - i\n",
        "            P[i] = min(limite - i, P[simetrica])\n",
        "\n",
        "        while (i + P[i] + 1 < e and\n",
        "               i - P[i] - 1 >= 0 and\n",
        "               texto[i + P[i] + 1] == texto[i - P[i] - 1]):\n",
        "            P[i] += 1\n",
        "\n",
        "        if i + P[i] > limite:\n",
        "            limite = i + P[i]\n",
        "            centro = i\n",
        "\n",
        "    maxVal = max(P)\n",
        "    maxIndex = P.index(maxVal)\n",
        "    inicio = (maxIndex - maxVal) // 2\n",
        "    fin = inicio + maxVal\n",
        "    return frase[inicio:fin], inicio, fin\n",
        "\n",
        "#Trie Trees para búsqeuda de aminoácidos en las proteinas\n",
        "class Nodo:\n",
        "    def __init__(self, valor):\n",
        "        self.valor = valor\n",
        "        self.isFinal = False\n",
        "        self.aminoacid = None\n",
        "        self.hijos = []\n",
        "\n",
        "    def agregar_hijo(self, valor):\n",
        "        nuevo_nodo = Nodo(valor)\n",
        "        self.hijos.append(nuevo_nodo)\n",
        "        return nuevo_nodo\n",
        "\n",
        "    def search_hijo(self, valor):\n",
        "        for hijo in self.hijos:\n",
        "            if hijo.valor == valor:\n",
        "                return hijo\n",
        "        return None\n",
        "\n",
        "    def __str__(self):\n",
        "        return str(self.valor)\n",
        "\n",
        "class Trie:\n",
        "    def __init__(self):\n",
        "        self.raiz = Nodo(\"#\")\n",
        "\n",
        "    def agregar_palabra(self, palabra, aminoacid):\n",
        "        actual = self.raiz\n",
        "        for letra in palabra:\n",
        "            hijo = actual.search_hijo(letra)\n",
        "            if hijo is None:\n",
        "                hijo = actual.agregar_hijo(letra)\n",
        "            actual = hijo\n",
        "        actual.isFinal = True\n",
        "        actual.aminoacid = aminoacid\n",
        "\n",
        "    def mostrar_arbol(self):\n",
        "        actual = self.raiz\n",
        "        queue = deque([actual])\n",
        "\n",
        "        while queue:\n",
        "            for _ in range(len(queue)):\n",
        "                nodo = queue.popleft()\n",
        "                print(nodo, end=\" \")\n",
        "                for hijo in nodo.hijos:\n",
        "                    queue.append(hijo)\n",
        "            print()\n",
        "\n",
        "    def buscar_palabra(self, palabra):\n",
        "        actual = self.raiz\n",
        "        for letra in palabra:\n",
        "            hijo = actual.search_hijo(letra)\n",
        "            if hijo is None:\n",
        "                return None\n",
        "            else:\n",
        "                actual = hijo\n",
        "        if actual.isFinal:\n",
        "            return actual.aminoacid\n",
        "        else:\n",
        "            return \"-\"\n",
        "\n",
        "def buscar_palabras(nodo, palabra):\n",
        "    palabras = []\n",
        "    if nodo.isFinal:\n",
        "        palabras.append(palabra)\n",
        "    for hijo in nodo.hijos:\n",
        "        palabras += buscar_palabras(hijo, palabra + hijo.valor)\n",
        "    return palabras\n",
        "\n",
        "# Buscar proteinas en secuencia\n",
        "def SearchProteinas(sequence):\n",
        "  lista_aminoacidos = []\n",
        "  reading_frames = [0,1,2]\n",
        "\n",
        "  # Por cada inicio de proteina, traducir hasta encontrar un STOP o hasta que acabe\n",
        "  while len(reading_frames) > 0:\n",
        "      start_index = reading_frames[0]\n",
        "      end_index = reading_frames[0]\n",
        "\n",
        "      protein = \"\"\n",
        "      for i in range(start_index, len(sequence), 3): # Itera de 3 en 3 para buscar los codones\n",
        "        codon = sequence[i:i+3]\n",
        "        if len(codon) == 3:\n",
        "          aminoacid = diccionario_aminoacidos.buscar_palabra(codon)\n",
        "          protein += aminoacid\n",
        "          end_index = i+3\n",
        "\n",
        "      lista_aminoacidos.append((reading_frames[0], protein))\n",
        "      reading_frames.pop(0)\n",
        "\n",
        "  results = []\n",
        "\n",
        "  # Buscar las proteinas en las listas encontradas\n",
        "  # for aminoacid in lista_aminoacidos:\n",
        "  #    for protein in proteins:\n",
        "  for protein in proteins:\n",
        "    encontrado = False\n",
        "    for aminoacid in lista_aminoacidos:\n",
        "      z = StringMatchZ(aminoacid[1], protein.sequence)\n",
        "      if len(z) > 0:\n",
        "        encontrado = True\n",
        "        results.append((protein.name, z[0]*3, protein.sequence[0:4], sequence[z[0]*3:z[0]*3+12]))\n",
        "    if not encontrado:\n",
        "      results.append((protein.name, \"No encontrado\", \"\", \"\"))\n",
        "  return results\n",
        "\n",
        "# Operaciones para el punto 4\n",
        "\n",
        "def get_codon(sequence, index):\n",
        "  return (diccionario_aminoacidos.buscar_palabra(sequence[index:index+3]), sequence[index:index+3])\n",
        "\n",
        "def get_diff(sequence1, sequence2):\n",
        "  results = []\n",
        "  reading_frames = [0,1]\n",
        "\n",
        "  # Por cada inicio de proteina, traducir hasta encontrar un STOP o hasta que acabe\n",
        "  while len(reading_frames) > 0:\n",
        "      start_index = reading_frames[0]\n",
        "      end_index = reading_frames[0]\n",
        "\n",
        "      protein = \"\"\n",
        "      for i in range(start_index, min(len(sequence1), len(sequence2)), 3): # Itera de 3 en 3 para buscar los codones\n",
        "        # codon1 = sequence1[i:i+3]\n",
        "        # codon2 = sequence2[i:i+3]\n",
        "        codon1, aminoacid = get_codon(sequence1, i)\n",
        "        codon2, aminoacid2 = get_codon(sequence2, i)\n",
        "        if codon1 != codon2:\n",
        "          aparece = False\n",
        "          for result in results:\n",
        "            if abs(result[0] - i) <= 20:\n",
        "              aparece = True\n",
        "              break\n",
        "          if not aparece:\n",
        "            results.append((i, (aminoacid, codon1), (aminoacid2, codon2)))\n",
        "\n",
        "      reading_frames.pop(0)\n",
        "\n",
        "  return results\n",
        "\n",
        "#Creación de diccionario Trie para búsqueda de aminoácidos\n",
        "\n",
        "aminoacids = [\n",
        "    # Alanine\n",
        "    (\"GCT\", \"A\"), (\"GCC\", \"A\"), (\"GCA\", \"A\"), (\"GCG\", \"A\"),\n",
        "    # Arginine\n",
        "    (\"CGT\", \"R\"), (\"CGC\", \"R\"), (\"CGA\", \"R\"), (\"CGG\", \"R\"), (\"AGA\", \"R\"), (\"AGG\", \"R\"),\n",
        "    # Asparagine\n",
        "    (\"AAT\", \"N\"), (\"AAC\", \"N\"),\n",
        "    # Aspartic Acid\n",
        "    (\"GAT\", \"D\"), (\"GAC\", \"D\"),\n",
        "    # Cysteine\n",
        "    (\"TGT\", \"C\"), (\"TGC\", \"C\"),\n",
        "    # Glutamic Acid\n",
        "    (\"GAA\", \"E\"), (\"GAG\", \"E\"),\n",
        "    # Glutamine\n",
        "    (\"CAA\", \"Q\"), (\"CAG\", \"Q\"),\n",
        "    # Glycine\n",
        "    (\"GGT\", \"G\"), (\"GGC\", \"G\"), (\"GGA\", \"G\"), (\"GGG\", \"G\"),\n",
        "    # Histidine\n",
        "    (\"CAT\", \"H\"), (\"CAC\", \"H\"),\n",
        "    # Isoleucine\n",
        "    (\"ATT\", \"I\"), (\"ATC\", \"I\"), (\"ATA\", \"I\"),\n",
        "    # Leucine\n",
        "    (\"CTT\", \"L\"), (\"CTC\", \"L\"), (\"CTA\", \"L\"), (\"CTG\", \"L\"), (\"TTA\", \"L\"), (\"TTG\", \"L\"),\n",
        "    # Lysine\n",
        "    (\"AAA\", \"K\"), (\"AAG\", \"K\"),\n",
        "    # Methionine (Start Codon)\n",
        "    (\"ATG\", \"M\"),\n",
        "    # Phenylalanine\n",
        "    (\"TTT\", \"F\"), (\"TTC\", \"F\"),\n",
        "    # Proline\n",
        "    (\"CCT\", \"P\"), (\"CCC\", \"P\"), (\"CCA\", \"P\"), (\"CCG\", \"P\"),\n",
        "    # Serine\n",
        "    (\"TCT\", \"S\"), (\"TCC\", \"S\"), (\"TCA\", \"S\"), (\"TCG\", \"S\"), (\"AGT\", \"S\"), (\"AGC\", \"S\"),\n",
        "    # Threonine\n",
        "    (\"ACT\", \"T\"), (\"ACC\", \"T\"), (\"ACA\", \"T\"), (\"ACG\", \"T\"),\n",
        "    # Tryptophan\n",
        "    (\"TGG\", \"W\"),\n",
        "    # Tyrosine\n",
        "    (\"TAT\", \"Y\"), (\"TAC\", \"Y\"),\n",
        "    # Valine\n",
        "    (\"GTT\", \"V\"), (\"GTC\", \"V\"), (\"GTA\", \"V\"), (\"GTG\", \"V\"),\n",
        "    # Stop Codons\n",
        "    (\"TAA\", \"*\"), (\"TAG\", \"*\"), (\"TGA\", \"*\")\n",
        "]\n",
        "\n",
        "amino_acid_dict = {\n",
        "    'A': 'Alanine',\n",
        "    'R': 'Arginine',\n",
        "    'N': 'Asparagine',\n",
        "    'D': 'Aspartic Acid',\n",
        "    'C': 'Cysteine',\n",
        "    'E': 'Glutamic Acid',\n",
        "    'Q': 'Glutamine',\n",
        "    'G': 'Glycine',\n",
        "    'H': 'Histidine',\n",
        "    'I': 'Isoleucine',\n",
        "    'L': 'Leucine',\n",
        "    'K': 'Lysine',\n",
        "    'M': 'Methionine',\n",
        "    'F': 'Phenylalanine',\n",
        "    'P': 'Proline',\n",
        "    'S': 'Serine',\n",
        "    'T': 'Threonine',\n",
        "    'W': 'Tryptophan',\n",
        "    'Y': 'Tyrosine',\n",
        "    'V': 'Valine',\n",
        "    '*': 'Stop Codon'\n",
        "}\n",
        "\n",
        "\n",
        "diccionario_aminoacidos = Trie()\n",
        "\n",
        "for aminoacido in aminoacids:\n",
        "    diccionario_aminoacidos.agregar_palabra(aminoacido[0], aminoacido[1])\n",
        "    print(aminoacido)\n"
      ],
      "metadata": {
        "id": "uUPMpr91c412",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "ea9ce730-978c-4db0-b781-4da6980c7ba7"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "('GCT', 'A')\n",
            "('GCC', 'A')\n",
            "('GCA', 'A')\n",
            "('GCG', 'A')\n",
            "('CGT', 'R')\n",
            "('CGC', 'R')\n",
            "('CGA', 'R')\n",
            "('CGG', 'R')\n",
            "('AGA', 'R')\n",
            "('AGG', 'R')\n",
            "('AAT', 'N')\n",
            "('AAC', 'N')\n",
            "('GAT', 'D')\n",
            "('GAC', 'D')\n",
            "('TGT', 'C')\n",
            "('TGC', 'C')\n",
            "('GAA', 'E')\n",
            "('GAG', 'E')\n",
            "('CAA', 'Q')\n",
            "('CAG', 'Q')\n",
            "('GGT', 'G')\n",
            "('GGC', 'G')\n",
            "('GGA', 'G')\n",
            "('GGG', 'G')\n",
            "('CAT', 'H')\n",
            "('CAC', 'H')\n",
            "('ATT', 'I')\n",
            "('ATC', 'I')\n",
            "('ATA', 'I')\n",
            "('CTT', 'L')\n",
            "('CTC', 'L')\n",
            "('CTA', 'L')\n",
            "('CTG', 'L')\n",
            "('TTA', 'L')\n",
            "('TTG', 'L')\n",
            "('AAA', 'K')\n",
            "('AAG', 'K')\n",
            "('ATG', 'M')\n",
            "('TTT', 'F')\n",
            "('TTC', 'F')\n",
            "('CCT', 'P')\n",
            "('CCC', 'P')\n",
            "('CCA', 'P')\n",
            "('CCG', 'P')\n",
            "('TCT', 'S')\n",
            "('TCC', 'S')\n",
            "('TCA', 'S')\n",
            "('TCG', 'S')\n",
            "('AGT', 'S')\n",
            "('AGC', 'S')\n",
            "('ACT', 'T')\n",
            "('ACC', 'T')\n",
            "('ACA', 'T')\n",
            "('ACG', 'T')\n",
            "('TGG', 'W')\n",
            "('TAT', 'Y')\n",
            "('TAC', 'Y')\n",
            "('GTT', 'V')\n",
            "('GTC', 'V')\n",
            "('GTA', 'V')\n",
            "('GTG', 'V')\n",
            "('TAA', '*')\n",
            "('TAG', '*')\n",
            "('TGA', '*')\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Lectura de archivos"
      ],
      "metadata": {
        "id": "CqA0vyeDzGEV"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "9KLn2JWeWxHh"
      },
      "outputs": [],
      "source": [
        "# Leer archivos\n",
        "\n",
        "with open('SARS-COV-2-MN908947.3.txt', 'r') as file:\n",
        "    lines = file.readlines()\n",
        "    secuencia_wuhan = \"\"\n",
        "    for line in lines[1:]:\n",
        "        line = line.strip()\n",
        "        secuencia_wuhan += line\n",
        "\n",
        "with open('SARS-COV-2-MT106054.1.txt', 'r') as file:\n",
        "    lines = file.readlines()\n",
        "    secuencia_texas = \"\"\n",
        "    for line in lines[1:]:\n",
        "        line = line.strip()\n",
        "        secuencia_texas += line\n",
        "\n",
        "with open('gen-M.txt', 'r') as file:\n",
        "    lines = file.readlines()\n",
        "    gen_m = \"\"\n",
        "    for line in lines[1:]:\n",
        "        line = line.strip()\n",
        "        gen_m += line\n",
        "\n",
        "with open('gen-S.txt', 'r') as file:\n",
        "    lines = file.readlines()\n",
        "    gen_s = \"\"\n",
        "    for line in lines[1:]:\n",
        "        line = line.strip()\n",
        "        gen_s += line\n",
        "\n",
        "with open('gen-ORF1AB.txt', 'r') as file:\n",
        "    lines = file.readlines()\n",
        "    gen_o = \"\"\n",
        "    for line in lines[1:]:\n",
        "        line = line.strip()\n",
        "        gen_o += line\n",
        "\n",
        "with open('seq-proteins.txt', 'r') as file:\n",
        "    lines = file.readlines()\n",
        "    proteins = []\n",
        "    for line in lines:\n",
        "        if line.startswith(\">\"):\n",
        "            name = line.strip()[1:]\n",
        "            sequence = \"\"\n",
        "        else:\n",
        "            sequence += line.strip()\n",
        "            proteins.append(Protein(name, sequence))"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 1. Aparicion de Los genes\n",
        "\n",
        "En la siguiente seccion se obtienen los **indices** de aparicion de los genes en la secuencia del virus *SARS-COV-2-MN908947.3*\n",
        "\n",
        "**Complejidades**\n",
        "\n",
        "Para la búsqueda de las cadenas de los genes en la cadena de Wuhan (2019) se utilizó string matching con Z-Array, que tiene una complejidad de O(n), donde en este caso, n es la cantidad de aminoácidos en la secuencia. El procedimiento se realiza una vez por cada gen a buscar."
      ],
      "metadata": {
        "id": "ZasD3DfMyjT_"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Indices de aparición de los genes en la secuencia del virus SARS-COV-2-MN908947.3.txt\n",
        "\n",
        "print(\"Apariciones del gen M: \")\n",
        "gen_m_indices = StringMatchZ(secuencia_wuhan, gen_m)\n",
        "print(gen_m_indices)\n",
        "print(secuencia_wuhan[gen_m_indices[0]:gen_m_indices[0]+12])\n",
        "\n",
        "print(\"Apariciones del gen S: \")\n",
        "gen_s_indices = StringMatchZ(secuencia_wuhan, gen_s)\n",
        "print(gen_s_indices)\n",
        "print(secuencia_wuhan[gen_s_indices[0]:gen_s_indices[0]+12])\n",
        "\n",
        "print(\"Apariciones del gen ORF1AB: \")\n",
        "gen_o_indices = StringMatchZ(secuencia_wuhan, gen_o)\n",
        "print(gen_o_indices)\n",
        "print(secuencia_wuhan[gen_o_indices[0]:gen_o_indices[0]+12])"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "W9-TqPqPZIQq",
        "outputId": "e80ab473-5d95-4566-d2bb-6f3f35d072ec"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Apariciones del gen M: \n",
            "[26522]\n",
            "ATGGCAGATTCC\n",
            "Apariciones del gen S: \n",
            "[21562]\n",
            "ATGTTTGTTTTT\n",
            "Apariciones del gen ORF1AB: \n",
            "[265]\n",
            "ATGGAGAGCCTT\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 2. Palindromo Mas largo\n",
        "\n",
        "En la siguiente seccion se muestran los palindromos mas largos para cada uno de los genomas\n",
        "\n",
        "\n",
        "---\n",
        "\n",
        "**Complejidades**\n",
        "Para encontrar palíndromos, se utilizó el algorítmo de Manacher, que cuenta con una complejidad O(n) donde n es la cantidad de aminoácidos en la secuencia."
      ],
      "metadata": {
        "id": "iE8eaba3KQMU"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Palíndromo más largo en cada uno de los tres genes\n",
        "\n",
        "print(\"Palíndromo más grande para gen M\")\n",
        "palindrome_m, inicio_m, fin_m = Manacher(gen_m)\n",
        "print(palindrome_m, \"\\tIndice:\", inicio_m, \"\\tLongitud:\", len(palindrome_m))\n",
        "print('\\n')\n",
        "\n",
        "print(\"Palíndromo más grande para gen S\")\n",
        "palindrome_s, inicio_s, fin_s = Manacher(gen_s)\n",
        "print(palindrome_s, \"\\tIndice:\", inicio_s, \"\\tLongitud:\", len(palindrome_s))\n",
        "print('\\n')\n",
        "\n",
        "print(\"Palíndromo más grande para gen ORF1AB\")\n",
        "palindrome_o, inicio_o, fin_o = Manacher(gen_o)\n",
        "print(palindrome_o, \"\\tIndice:\", inicio_o, \"\\tLongitud:\", len(palindrome_o))\n",
        "print('\\n')\n",
        "\n",
        "# Escribir palindromos en archivo\n",
        "with open('palindromos.txt', 'w') as file:\n",
        "  file.write(\"M\\n\"+str(inicio_m)+\",\"+str(fin_m)+\",\"+palindrome_m+\"\\n\")\n",
        "  file.write(\"S\\n\"+str(inicio_s)+\",\"+str(fin_s)+\",\"+palindrome_s+\"\\n\")\n",
        "  file.write(\"ORF1AB\\n\"+str(inicio_o)+\",\"+str(fin_o)+\",\"+palindrome_o+\"\\n\")\n",
        "  file.close()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "t59tHiECHZC2",
        "outputId": "898c96f3-4984-4800-d011-afa882eac98e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Palíndromo más grande para gen M\n",
            "CTAAAGAAATC \tIndice: 493 \tLongitud: 11\n",
            "\n",
            "\n",
            "Palíndromo más grande para gen S\n",
            "GAATTTGTGTTTAAG \tIndice: 570 \tLongitud: 15\n",
            "\n",
            "\n",
            "Palíndromo más grande para gen ORF1AB\n",
            "CTCAATGACTTCAGTAACTC \tIndice: 9711 \tLongitud: 20\n",
            "\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 3. Busqueda de Proteinas\n",
        "\n",
        "En este apartado mostramos las apariciones de las proteinas de \"seq-proteins.txt\" en la secuencia del virus de de Wuhan (2019)\n",
        "\n",
        "Se muestra el nombre de la proteína, el índice donde se comienza a producir, sus primeros aminoácidos y sus codones asociados\n",
        "\n",
        "---\n",
        "\n",
        "**Estructuras**\n",
        "\n",
        "Para relacionar trios de aminoácidos con codones, se utilizó un diccionario Trie que ayuda a almacenar y encontrar las diferentes posibilidades de codones en una sola estructura.\n",
        "\n",
        "**Procedimiento**\n",
        "\n",
        "Se encontraron los posibles reading frames para la lectura y traducción. A partir de ahí, se leen los codones de 3 en 3 aminoácidos hasta acabar la secuencia. Una vez traducidos los reading frames, se buscan las proteínas en la secuencia generada con String Matching usando un Z array.\n",
        "\n",
        "**Complejidad**\n",
        "\n",
        " El procedimiento completo tiene una complejidad en el dominio de O(n), donde n es la cantidad de aminoácidos en la secuencia."
      ],
      "metadata": {
        "id": "Mc68-yK0ToiP"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "res = SearchProteinas(secuencia_wuhan)\n",
        "for re in res:\n",
        "  if re[1] != \"No encontrado\":\n",
        "    print(re[0], \":\", re[1], \"- 4 amino:\", re[2], \"=\", re[3])\n",
        "  else:\n",
        "    print(re[0], \":\", re[1])"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "dugzafPWbYI-",
        "outputId": "fd992051-9ca3-47c8-d6d2-935883c744be"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "QHD43415_1 : 264 - 4 amino: MESL = GATGGAGAGCCT\n",
            "QHD43415_2 : 804 - 4 amino: AYTR = GGCATACACTCG\n",
            "QHD43415_3 : 2718 - 4 amino: APTK = TGCACCAACAAA\n",
            "QHD43415_4 : 8553 - 4 amino: KIVN = TAAAATTGTTAA\n",
            "QHD43415_5 : 10053 - 4 amino: SGFR = GAGTGGTTTTAG\n",
            "QHD43415_6 : 10971 - 4 amino: SAVK = AAGTGCAGTGAA\n",
            "QHD43415_7 : 11841 - 4 amino: SKMS = GTCTAAAATGTC\n",
            "QHD43415_8 : 12090 - 4 amino: AIAS = AGCTATAGCCTC\n",
            "QHD43415_9 : 12684 - 4 amino: NNEL = GAATAATGAGCT\n",
            "QHD43415_10 : 13023 - 4 amino: AGNA = AGCTGGTAATGC\n",
            "QHD43415_11 : No encontrado\n",
            "QHD43415_12 : 16236 - 4 amino: AVGA = GCTGTTGGGGCT\n",
            "QHD43415_13 : 18039 - 4 amino: AENV = GCTGAAAATGTA\n",
            "QHD43415_14 : 19620 - 4 amino: SLEN = AGTTTAGAAAAT\n",
            "QHD43415_15 : 20658 - 4 amino: SSQA = TCTAGTCAAGCG\n",
            "QHD43416 : 21561 - 4 amino: MFVF = AATGTTTGTTTT\n",
            "QHD43417 : 25392 - 4 amino: MDLF = ATGGATTTGTTT\n",
            "QHD43418 : 26244 - 4 amino: MYSF = ATGTACTCATTC\n",
            "QHD43419 : 26520 - 4 amino: MADS = CCATGGCAGATT\n",
            "QHD43420 : 27201 - 4 amino: MFHL = ATGTTTCATCTC\n",
            "QHD43421 : 27393 - 4 amino: MKII = ATGAAAATTATT\n",
            "QHD43422 : 27891 - 4 amino: MKFL = ACATGAAATTTC\n",
            "QHD43423 : 28272 - 4 amino: MSDN = AATGTCTGATAA\n",
            "QHI42199 : 29556 - 4 amino: MGYI = GATGGGCTATAT\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 4. Comparar proteinas entre Wuhan (2019) y Texas (2020)\n",
        "\n",
        "Para este punto, similar al punto 3, se calcularon y se tradujeron los reading frames para realizar comparación entre ambas secuencias desde las diferentes perspectivas de los reading frames, respuestas repetidas no se tomaron en cuenta.\n",
        "\n",
        "**Complejidad**\n",
        "\n",
        "Se recorren 3 veces las secuencias, por lo que el algoritmo tiene una complejidad en el dominio de O(n), donde n es la cantidad de aminoácidos en la secuencia."
      ],
      "metadata": {
        "id": "OfFpES20gbDI"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "res = get_diff(secuencia_wuhan, secuencia_texas)\n",
        "print(\"Indice\\t\\tWuhan\\t\\tTexas\")\n",
        "for r in res:\n",
        "  # print(r)\n",
        "  # print(\"Diferencia en\", r[0], \":\", r[1][0], \"(\"+r[1][1]+\")\", \"vs\", r[2][0], \"(\"+r[2][1]+\")\")\n",
        "  print(str(r[0]) + \"\\t\\t\" + str(r[1][0])+ \" (\" + str(r[1][1]) + \")\" + \"\\t\\t\" + str(r[2][0])+ \" (\" + str(r[2][1]) + \")\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "76_ggyrYsfEP",
        "outputId": "498ef9b6-84c0-4aa7-e8e7-56dd6c26d35a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Indice\t\tWuhan\t\tTexas\n",
            "8781\t\tCCA (P)\t\tTCA (S)\n",
            "19173\t\tGAT (D)\t\tGCT (A)\n",
            "27924\t\tCAA (Q)\t\tTAA (*)\n",
            "28143\t\tTAC (Y)\t\tCAC (H)\n",
            "29094\t\tCGG (R)\t\tTGG (W)\n",
            "29880\t\tAAA (K)\t\tAA (-)\n",
            "18601\t\tATG (M)\t\tACG (T)\n",
            "18973\t\tTTC (F)\t\tTAC (Y)\n"
          ]
        }
      ]
    }
  ]
}